blocks/algorithms/__init__.py

```diff
@@ -234,12 +247,35 @@ def initialize(self):
         self._function = theano.function(self.inputs, [], updates=all_updates)
         logger.info("The training algorithm is initialized")
 
-    def process_batch(self, batch):
-        if not set(batch.keys()) == set([v.name for v in self.inputs]):
-            raise ValueError("mismatch of variable names and data sources" +
-                             variable_mismatch_error.format(
+    def _validate_source_names(self, batch):
+        in_names = [v.name for v in self.inputs]
+
+        if not set(in_names).issubset(set(batch.keys())):
+            raise ValueError("Didn't find all sources: " +
+                             source_missing_error.format(
                                  sources=batch.keys(),
-                                 variables=[v.name for v in self.inputs]))
+                                 variables=in_names))
+        if not set(batch.keys()).issubset(set(in_names)):
+            if self.on_unused_sources == 'ignore':
+                pass
+            elif self.on_unused_sources == 'warn':
+                if not hasattr(self, '_unused_sources_warning'):
+                    logger.warn(variable_mismatch_error.format(
+                        sources=batch.keys(),
+                        variables=in_names))
+                self._unused_sources_warning = True
```

Comment:
Naming: maybe `_unused_source_warned`? Right now it seems like it is name of a flag, which it is not.